dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/JdbcTrackedEntityChangeLogStore.java

```diff
@@ -0,0 +1,112 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.export.trackedentity;
+
+import java.util.List;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.hisp.dhis.program.UserInfoSnapshot;
+import org.hisp.dhis.tracker.export.trackedentity.TrackedEntityChangeLog.Change;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.core.namedparam.SqlParameterSource;
+import org.springframework.stereotype.Repository;
+
+@Repository("org.hisp.dhis.tracker.export.trackedentity.JdbcTrackedEntityChangeLogStore")
+@RequiredArgsConstructor
+public class JdbcTrackedEntityChangeLogStore {
+
+  private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+  private static final RowMapper<TrackedEntityChangeLog> customTrackedEntityChangeLogRowMapper =
+      (rs, rowNum) -> {
+        UserInfoSnapshot createdBy = new UserInfoSnapshot();
+        createdBy.setUsername(rs.getString("userName"));
+        createdBy.setFirstName(rs.getString("firstname"));
+        createdBy.setSurname(rs.getString("surname"));
+        createdBy.setUid(rs.getString("useruid"));
+
+        return new TrackedEntityChangeLog(
+            createdBy,
+            rs.getTimestamp("updatedAt"),
+            rs.getString("type"),
+            new Change(
+                new TrackedEntityChangeLog.TrackedEntityAttributeChange(
+                    rs.getString("trackedEntityAttributeUid"),
+                    rs.getString("previousValue"),
+                    rs.getString("currentValue"))));
+      };
+
+  public List<TrackedEntityChangeLog> getTrackedEntityChangeLog(
+      String trackedEntity, Set<String> attributes) {
```

Comment:
```suggestion
      UID trackedEntity, Set<String> attributes) {
```

We should use our `UID` type where we can. We lose information if we move from UID in the service to a String here.

Reply:
We lose type information if we move from UID to a String. Why would we do that?

The reason we are using it in the web is because it all started with OpenAPI and then conversion of controller parameters to UID.